### PR TITLE
L2 storage framework

### PR DESCRIPTION
**Pull request by bobotu**
The basic framework to support store SST on level 2 storage.

With some modification on ristretto cache https://github.com/dgraph-io/ristretto/compare/master...bobotu:master?expand=1

### CODE REVIEW

File: l2/storage.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	return diskStorage{
+		path: path,
+	}, nil
+}
+
+func (s diskStorage) Put(path, name string) error {
+	src := filepath.Join(path, name)
+	dst := filepath.Join(s.path, name)
+	return os.Link(src, dst)
```
Comment (coocood) [inline l2/storage.go]: I think we should copy the file.

File: l2/storage.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+func (s diskStorage) Get(name, path string) error {
+	src := filepath.Join(s.path, name)
+	dst := filepath.Join(path, name)
+	return os.Link(src, dst)
```
Comment (coocood) [inline l2/storage.go]: ditto

File: l2/storage.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package l2
+
+import (
+	"os"
+	"path/filepath"
+)
+
+type storage interface {
+	Put(path, name string) error
+	Get(name, path string) error
```
Comment (coocood) [inline l2/storage.go]: I think we should be able to Put/Get a fraction of a file instead of the whole file.

Comment (bobotu) [reply l2/storage.go]: This is an optimization that can be adopted in later PR.

File: l2/l2.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+const savingSuffix = ".saving"
+
+type Storage struct {
+	cache       *ristretto.Cache
+	resourceMgr *epoch.ResourceManager
+	storage     storage
+	opts        *Options
+}
+
+func NewCache(opts *Options, resourceMgr *epoch.ResourceManager) (*Storage, error) {
```
Comment (coocood) [inline l2/l2.go]: NewStorage

File: levels.go [hunk 5]
```diff
@@ -116,7 +116,15 @@ func newLevelsController(kv *DB, mf *Manifest, mgr *epoch.ResourceManager, opt o
...
 		}
 
-		t, err := table.OpenTable(fname, tableManifest.Compression, kv.blockCache)
+		cfg := &table.OpenTableConfig{
+			ID:          fileID,
+			Path:        kv.opt.Dir,
+			Compression: tableManifest.Compression,
+			Cache:       kv.blockCache,
+			UseL2:       int(tableManifest.Level) >= kv.opt.L2StartLevel,
+			L2Cache:     kv.l2,
```
Comment (coocood) [inline levels.go]: L2Storage

**Pull request closed without merging**